File: pkginstaller.py
#!/usr/bin/env python3
import subprocess
import os
import sys
from typing import List, Tuple
import time
import fcntl
import re

PACKAGE_FILE = "packages.txt"

def strip_version(package_line: str) -> str:
    """Strip version number from package name"""
    # Match package name before any version number or space
    match = re.match(r'^([a-zA-Z0-9@_+-]+)', package_line)
    if match:
        return match.group(1)
    return package_line.split()[0]  # Fallback: just take first word

def is_package_installed(package_name: str) -> bool:
    """Check if a package is installed"""
    try:
        result = subprocess.run(
            ["pacman", "-Q", package_name],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            check=False
        )
        return result.returncode == 0
    except Exception as e:
        print(f"Error checking package {package_name}: {str(e)}")
        return False

def install_package(package_name: str) -> bool:
    """Install a package with proper locking"""
    lock_file = "/tmp/pacman_install.lock"
    
    try:
        with open(lock_file, 'w') as f:
            try:
                fcntl.flock(f.fileno(), fcntl.LOCK_EX)
                
                # Strip version number if present
                package_name = strip_version(package_name)
                
                result = subprocess.run(
                    ["sudo", "pacman", "-S", "--noconfirm", package_name],
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    text=True,
                    check=False
                )
                
                if result.returncode != 0:
                    print(f"Failed to install {package_name}: {result.stderr.strip()}")
                    return False
                    
                return True
                
            finally:
                fcntl.flock(f.fileno(), fcntl.LOCK_UN)
                
    except Exception as e:
        print(f"Error installing package {package_name}: {str(e)}")
        return False

def main():
    if os.geteuid() == 0:
        print("Don't run this script as root! It will ask for sudo when needed.")
        sys.exit(1)

    try:
        subprocess.run(["sudo", "-v"], check=True)
    except subprocess.CalledProcessError:
        print("Failed to verify sudo access. Please make sure you have sudo privileges.")
        sys.exit(1)

    if not os.path.exists(PACKAGE_FILE):
        print(f"Package list file '{PACKAGE_FILE}' not found.")
        sys.exit(1)

    try:
        with open(PACKAGE_FILE, "r") as file:
            packages = [line.strip() for line in file if line.strip() and not line.startswith('#')]
    except Exception as e:
        print(f"Error reading package file: {e}")
        sys.exit(1)

    print(f"Found {len(packages)} packages to process")
    
    failed_packages = []
    successful_packages = []
    already_installed = []

    for package_line in packages:
        package_name = strip_version(package_line)
        print(f"\nProcessing package: {package_name}")
        
        if is_package_installed(package_name):
            print(f"✓ {package_name} is already installed")
            already_installed.append(package_name)
            continue
            
        print(f"Installing {package_name}...")
        
        subprocess.run(["sudo", "-v"], check=False)
        
        if install_package(package_name):
            print(f"✓ Successfully installed {package_name}")
            successful_packages.append(package_name)
        else:
            print(f"✗ Failed to install {package_name}")
            failed_packages.append(package_name)
        
        time.sleep(0.5)

    print("\n" + "="*50)
    print("Installation Summary:")
    print(f"Total packages processed: {len(packages)}")
    print(f"Already installed: {len(already_installed)}")
    print(f"Successfully installed: {len(successful_packages)}")
    print(f"Failed installations: {len(failed_packages)}")

    if failed_packages:
        print("\nFailed packages:")
        for pkg in failed_packages:
            print(f"- {pkg}")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\nProcess interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        sys.exit(1)

File: pkginstaller.rs
use colored::*;
use rayon::prelude::*;
use std::{
    fs::File,
    io::{self, BufRead, BufReader},
    path::Path,
    process::Command,
    sync::{Arc, Mutex},
};

#[derive(Debug)]
struct PackageResult {
    name: String,
    status: InstallStatus,
}

#[derive(Debug)]
enum InstallStatus {
    AlreadyInstalled,
    Installed,
    Failed(String),
}

// Using a single sudo authentication for the entire session
fn ensure_sudo_session() -> Result<(), String> {
    let output = Command::new("sudo")
        .args(["-v"]) // Validate and update timestamp
        .output()
        .map_err(|e| e.to_string())?;

    if !output.status.success() {
        return Err("Failed to validate sudo privileges".to_string());
    }
    Ok(())
}

fn is_package_installed(package: &str) -> bool {
    Command::new("pacman")
        .args(["-Q", package])
        .output()
        .map(|output| output.status.success())
        .unwrap_or(false)
}

fn install_package(package: &str) -> Result<(), String> {
    // Using sudo directly since we validated the session earlier
    let output = Command::new("sudo")
        .args(["pacman", "-S", "--noconfirm", package])
        .output()
        .map_err(|e| e.to_string())?;

    if output.status.success() {
        Ok(())
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr);
        // Check for common pacman errors
        if stderr.contains("could not be found") {
            Err(format!("Package '{}' not found in repositories", package))
        } else if stderr.contains("could not satisfy dependencies") {
            Err("Dependency resolution failed".to_string())
        } else {
            Err(stderr.to_string())
        }
    }
}

fn process_package(package: &str) -> PackageResult {
    let name = package.to_string();

    // Print start of processing with lock
    {
        println!("Processing package: {}", name);
    }

    if is_package_installed(package) {
        return PackageResult {
            name,
            status: InstallStatus::AlreadyInstalled,
        };
    }

    // Keep sudo session alive
    let _ = Command::new("sudo").args(["-v"]).output();

    let status = match install_package(package) {
        Ok(_) => InstallStatus::Installed,
        Err(e) => InstallStatus::Failed(e),
    };

    PackageResult { name, status }
}

fn read_packages<P: AsRef<Path>>(path: P) -> io::Result<Vec<String>> {
    let file = File::open(path)?;
    let reader = BufReader::new(file);
    let packages: Vec<String> = reader
        .lines()
        .filter_map(|line| {
            let line = line.ok()?;
            let trimmed = line.trim();
            if !trimmed.is_empty() && !trimmed.starts_with('#') {
                Some(trimmed.to_string())
            } else {
                None
            }
        })
        .collect();
    Ok(packages)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Check if running as root
    if !nix::unistd::Uid::effective().is_root() {
        eprintln!("{}", "This script must be run with sudo privileges.".red());
        std::process::exit(1);
    }

    // Ensure sudo session is active
    if let Err(e) = ensure_sudo_session() {
        eprintln!("{}: {}", "Failed to establish sudo session".red(), e);
        std::process::exit(1);
    }

    // Read package list
    let packages = match read_packages("packages.txt") {
        Ok(pkgs) => pkgs,
        Err(e) => {
            eprintln!("{}: {}", "Error reading packages.txt".red(), e);
            std::process::exit(1);
        }
    };

    println!("Found {} packages to process", packages.len());

    // Track results
    let results = Arc::new(Mutex::new(Vec::new()));
    let results_clone = Arc::clone(&results);

    // Process packages in parallel with a controlled thread pool
    rayon::ThreadPoolBuilder::new()
        .num_threads(4) // Limit concurrent installations
        .build_global()
        .unwrap();

    packages.par_iter().for_each(|package| {
        let result = process_package(package);

        // Print progress with lock to avoid garbled output
        {
            let status_str = match &result.status {
                InstallStatus::AlreadyInstalled => {
                    format!("{} {} is already installed", "✓".green(), package).green()
                }
                InstallStatus::Installed => {
                    format!("{} Successfully installed {}", "✓".green(), package).green()
                }
                InstallStatus::Failed(err) => {
                    format!("{} Failed to install {}: {}", "✗".red(), package, err).red()
                }
            };
            println!("{}", status_str);
        }

        results_clone.lock().unwrap().push(result);
    });

    // Print summary
    let results = results.lock().unwrap();
    let total = results.len();
    let (success, already, failed): (Vec<_>, Vec<_>, Vec<_>) = results
        .iter()
        .partition3(|r| matches!(r.status, InstallStatus::Installed));

    println!("\n{}", "=".repeat(50));
    println!("Installation Summary:");
    println!("Total packages processed: {}", total);
    println!("Already installed: {}", already.len());
    println!("Successfully installed: {}", success.len());
    println!("Failed installations: {}", failed.len());

    if !failed.is_empty() {
        println!("\nFailed packages:");
        for result in failed {
            if let InstallStatus::Failed(err) = &result.status {
                println!("- {}: {}", result.name.red(), err);
            }
        }
    }

    Ok(())
}
